LLVM 6 migration

amberhu-uw opened the pull request.

amberhu-uw committed:
+LLVM 6 migration

amberhu-uw committed:
+LLVM 6 migration

amberhu-uw committed:
=update Travis LLVM

amberhu-uw committed:
=update Travis LLVM

amberhu-uw committed:
=does LLVM 5 work?

amberhu-uw committed:
=does LLVM 5 work?

amberhu-uw committed:
=distro: xenial

amberhu-uw committed:
=distro: xenial

amberhu-uw committed:
=LLVM min ver 6

amberhu-uw committed:
=LLVM min ver 6

amberhu-uw committed:
=CMake find LLVM 6

amberhu-uw committed:
=CMake find LLVM 6

amberhu-uw committed:
=CMake don't pass wrong LLVM-config

amberhu-uw committed:
=CMake don't pass wrong LLVM-config

amberhu-uw committed:
=use C++11

amberhu-uw committed:
=use C++11

amberhu-uw committed:
=travis debugging

amberhu-uw committed:
=travis debugging

amberhu-uw committed:
=travis debugging

amberhu-uw committed:
=travis debugging

amberhu-uw committed:
=use c++14

amberhu-uw committed:
=use c++14

amberhu-uw committed:
=RR_USE_CXX14 in src

amberhu-uw committed:
=RR_USE_CXX14 in src

amberhu-uw committed:
=RR_USE_CXX14 in src

amberhu-uw committed:
=RR_USE_CXX14 in src

amberhu-uw committed:
=g++6

amberhu-uw committed:
=g++6

amberhu-uw committed:
=update yaml

amberhu-uw committed:
=update yaml

amberhu-uw committed:
=valid yaml syntax

amberhu-uw committed:
=valid yaml syntax

amberhu-uw committed:
=export g++-6

amberhu-uw committed:
=export g++-6

amberhu-uw committed:
=install g++-6

amberhu-uw committed:
=install g++-6

amberhu-uw committed:
=use g++

amberhu-uw committed:
=use g++

amberhu-uw committed:
=g++-6 addon

amberhu-uw committed:
=g++-6 addon

amberhu-uw committed:
=link g++-6 in

amberhu-uw committed:
=link g++-6 in

amberhu-uw committed:
=wrong place

amberhu-uw committed:
=wrong place

amberhu-uw committed:
=rename to g++

amberhu-uw committed:
=rename to g++

amberhu-uw committed:
=artful please

amberhu-uw committed:
=artful please

amberhu-uw committed:
=bionic?

amberhu-uw committed:
=bionic?

amberhu-uw committed:
-make_unique :(

amberhu-uw committed:
-make_unique :(

amberhu-uw committed:
=function pointer cast

amberhu-uw committed:
=function pointer cast

0u812 committed:
Update deps via apt

Hi Andrew, let's see if this fixes the problem

0u812 committed:
Update README.md

0u812 committed:
Update README.md

0u812 committed:
Update .travis.yml

0u812 committed:
Update .travis.yml

0u812 committed:
Update .travis.yml

0u812 committed:
update deps

0u812 committed:
Update .travis.yml

amberhu-uw committed:
=update travis

amberhu-uw committed:
=update travis

amberhu-uw committed:
Merge branch 'develop' into llvm-6

0u812 committed:
Update roadrunner.i

0u812 committed:
Update roadrunner.i

amberhu-uw committed:
=println debug

amberhu-uw committed:
=println debug

amberhu-uw committed:
Merge branch 'llvm-6' of https://github.com/sys-bio/roadrunner into llvm-6

0u812 committed:
debugging llvm installation

0u812 committed:
select llvm 6 config

0u812 committed:
select llvm 6 config

0u812 committed:
Update .travis.yml

0u812 committed:
Update .travis.yml

0u812 committed:
try adding llvm dir before

0u812 committed:
try adding llvm dir before

amberhu-uw committed:
=notice log level

amberhu-uw committed:
I am stupid Merge branch 'llvm-6' of https://github.com/sys-bio/roadrunner into llvm-6

0u812 committed:
debugging logs

0u812 committed:
debugging logs

0u812 committed:
Update LLVMExecutableModel.cpp

0u812 committed:
Update LLVMExecutableModel.cpp

amberhu-uw committed:
+getFunctionAddress & fix symbol name misspellings

amberhu-uw committed:
+getFunctionAddress & fix symbol name misspellings

amberhu-uw committed:
Merge branch 'llvm-6' of https://github.com/sys-bio/roadrunner into llvm-6

amberhu-uw committed:
-not verbose

amberhu-uw committed:
-not verbose

amberhu-uw committed:
-only one verbose

amberhu-uw committed:
-only one verbose

amberhu-uw committed:
-get rid of diagnostic print

amberhu-uw committed:
-get rid of diagnostic print

amberhu-uw committed:
-set a couple logs to debug

amberhu-uw committed:
-set a couple logs to debug

amberhu-uw committed:
-benchmark failing

amberhu-uw committed:
-benchmark failing

amberhu-uw committed:
-benchmark failing

amberhu-uw committed:
-benchmark failing

amberhu-uw committed:
Merge branch 'develop' into llvm-6

amberhu-uw committed:
=remove logging, speed up JIT

amberhu-uw committed:
=remove logging, speed up JIT

amberhu-uw committed:
Merge branch 'develop' into llvm-6

amberhu-uw committed:
Attempt to fix Travis

amberhu-uw committed:
Attempt to fix Swig source not found

amberhu-uw committed:
Find out where swig is

amberhu-uw committed:
Where are the swig libraries?

amberhu-uw committed:
Change SWIG_DIR

amberhu-uw committed:
Change SWIG_DIR

amberhu-uw committed:
Where are the debs being installed?

amberhu-uw committed:
Do I need sudo?

amberhu-uw committed:
Get debs package name

amberhu-uw committed:
Get package info

amberhu-uw committed:
see dpkg libroadrunner-deps-dev

amberhu-uw committed:
Try without TR1 CXX

amberhu-uw committed:
Use C++11

amberhu-uw committed:
C++14 instead of C++11

amberhu-uw committed:
C++14 instead of C++11

nobody committed:
Fixed segfault in rrSparse

nobody committed:
Fixed segfault in rrSparse

amberjhu committed:
Merge remote-tracking branch 'origin/llvm-6' into llvm-6

I literally have no idea if this is the right thing to do, why is rebasing so hard

amberjhu committed:
=LibFunc::fmod -> LibFunc_fmod (from Kiri's rebased commits)

amberjhu commented on source/llvm/CodeGenBase.h:
```diff
@@ -83,7 +83,7 @@ class CodeGenBase
      * function pass manager. Null if no optimization.
      */
-    llvm::FunctionPassManager *functionPassManager;
+    llvm::legacy::FunctionPassManager *functionPassManager;
```
This is deprecated, it's fine to use this here, but add something in the comment about it being legacy and what we're going to replace it with in the future when legacy FunctionPassManager is removed

amberjhu commented on source/llvm/LLVMModelGenerator.cpp:
```diff
@@ -119,120 +120,253 @@ ExecutableModel* LLVMModelGenerator::createModel(const std::string& sbml,
-
-        rc->setFloatingSpeciesConcentrationPtr =
-                SetFloatingSpeciesConcentrationCodeGen(context).createFunction();
+	Function* setBoundarySpeciesAmountIR = 0 ;
```
Using `nullptr` is better style instead of 0. I know that you need C++11, which is why the existing code hasn't been using that style, but we should start moving to it now that we require C++11 for `unique_ptr` and such

amberjhu commented on source/llvm/LLVMModelGenerator.cpp:
```diff
@@ -119,120 +120,253 @@ ExecutableModel* LLVMModelGenerator::createModel(const std::string& sbml,
+		setGlobalParameterInitValueIR			= 0;
+	}
+
+	//https://stackoverflow.com/questions/28851646/llvm-jit-windows-8-1
```
Can you explain this with more than just a link, so it's clear why we're doing this?

amberjhu commented on source/llvm/LLVMModelGenerator.cpp:
```diff
@@ -119,120 +120,253 @@ ExecutableModel* LLVMModelGenerator::createModel(const std::string& sbml,
+	context.getExecutionEngine().finalizeObject();
+
+	rc->evalInitialConditionsPtr = (EvalInitialConditionsCodeGen::FunctionPtr)
+        context.getExecutionEngine().getFunctionAddress("evalInitialConditions");
```
Does getFunctionAddress() do the compilation or is that finalizeObject()?

amberjhu commented on source/llvm/ModelDataIRBuilder.cpp:
```diff
@@ -106,7 +106,10 @@ llvm::Value* ModelDataIRBuilder::createGEP(ModelDataFields field,
+	// FIXME:
+	// The interface for this function was changed, and I'm unsure how to get the correct type, so I just made it double
+	// Okay, now I think I understand how to get the actual type
+    return builder.CreateStructGEP(cast<PointerType>(modelData->getType()->getScalarType())->getElementType(), modelData, (unsigned)field, Twine(fieldName) + Twine("_gep"));
```
Remove the FIXME and maybe explain what you now know about GEP and why the interface changed

amberjhu commented on source/llvm/ModelDataIRBuilder.cpp:
```diff
@@ -199,7 +205,7 @@ llvm::Function* ModelDataIRBuilder::getCSRMatrixSetNZDecl(Module *module)
                 IntegerType::get(module->getContext(), sizeof(bool) * 8), args,
                 false);
-        f = Function::Create(funcType, Function::InternalLinkage,
+        f = Function::Create(funcType, Function::ExternalLinkage,
```
Why did this change from external to internal linkage? What do those mean?

amberjhu commented on source/llvm/ModelGeneratorContext.h:
```diff
@@ -114,7 +115,7 @@ class ModelGeneratorContext
      * NULL if no optimization is specified.
      */
-    llvm::FunctionPassManager *getFunctionPassManager() const;
+    llvm::legacy::FunctionPassManager *getFunctionPassManager() const;
```
Again, add a message about what we will do about this in the future. Maybe a FIXME

amberjhu commented on source/llvm/ModelGeneratorContext.h:
```diff
@@ -162,7 +163,8 @@ class ModelGeneratorContext
      * that are accessible from the LLVM generated code.
      */
-    void addGlobalMappings();
+	void addGlobalMapping(const llvm::GlobalValue * gv, void * addr);
```
Why are there two functions now? Can you add two sets of documentation so we know when to call each of these?

amberjhu commented on source/llvm/Random.cpp:
```diff
@@ -119,6 +133,21 @@ void addGlobalMappings(const ModelGeneratorContext& ctx)
                     FunctionType::get(double_type, args_void_double_double, false), module),
                         (void*) distrib_normal);
+    
+    executionEngine->addGlobalMapping(
```
Isn't there an `addGlobalMapping` function now? Shouldn't we be using that?

amberjhu commented on source/llvm_testing/src/tests.cpp:
```diff
@@ -84,7 +84,7 @@ bool runInitialValueAssigmentTest(const string& version, int caseNumber)
         EvalInitialConditionsCodeGen::FunctionPtr pfunc;
 
-        pfunc = iv.createFunction();
+		pfunc = (EvalInitialConditionsCodeGen::FunctionPtr) engine.getPointerToFunction(iv.createFunction());
```
Why did we use `getFunctionAddress` earlier and now we're using `getPointerToFunction`?

amberjhu reviewed (commented):
I need to add this documentation, but otherwise looks good!

faye-u committed:
Add documentaton in ModelData

Add documentation in ModelDat

faye-u commented on source/llvm/LLVMExecutableModel.cpp:
```diff
@@ -138,6 +138,10 @@ int LLVMExecutableModel::setValues(bool (*funcPtr)(LLVMModelData*, int, double),
         bool result =  funcPtr(modelData, j, values[i]);
+        {
+          string id = (this->*getNameFuncPtr)(j);
+          Log(Logger::LOG_ERROR) << "LLVMExecutableModel::setValues for " << id << ", has initial assignment rule: " << symbols->hasInitialAssignmentRule(id);
```
Why use LOG_ERROR rather than LOG_INFORMATION?

amberjhu replied:
I believe this is outdated and is no longer in `llvm-6`. Check out that [same line](https://github.com/sys-bio/roadrunner/blob/llvm-6/source/llvm/LLVMExecutableModel.cpp#L141) from the current `llvm-6` branch and the log statement is not there.

mengfy0718 commented on wrappers/C/Testing/Suite_TestModel.cpp:
```diff
@@ -1678,7 +1678,7 @@ SUITE(TEST_MODEL)
         if(!values || values->Count != refList.size())
         {
-            CHECK(false);
+            CHECK(false && " ");
```
Is it the same as CHECK(false)?

faye-u commented on source/NLEQ1Solver.cpp:
```diff
@@ -58,10 +58,10 @@ void NLEQ1Solver::resetSettings()
 
     // Set default settings.
-    addSetting("allow_presimulation", true, "Allow Presimulation", "Flag for starting steady state analysis with simulation (bool).", "(bool) This flag does not affect the usage of approximation routine when the default steaty state solver fails");
+    addSetting("allow_presimulation", false, "Allow Presimulation", "Flag for starting steady state analysis with simulation (bool).", "(bool) This flag does not affect the usage of approximation routine when the default steaty state solver fails");
```
Why change the flag of "allow_presimulation" and "allow_approx" to false?

faye-u commented on source/llvm/LLVMExecutableModel.cpp:
```diff
@@ -730,7 +732,7 @@ void LLVMExecutableModel::reset(int opt)
             if (dirty_cm)
             {
-                Log(Logger::LOG_WARNING) << "Both initial conditions and "
+                Log(Logger::LOG_ERROR) << "Both initial conditions and "
```
Why use LOG_ERROR rather than LOG_WARNING?

faye-u commented on wrappers/C/Testing/Suite_TestModel.cpp:
```diff
@@ -1684,7 +1684,7 @@ SUITE(TEST_MODEL)
         if(!values || values->Count != refList.size())
         {
-            CHECK(false);
+            CHECK(false && " ");
```
I thought C++ uses short-circuit logic, so what's the difference between false and false&&" "?

andrewbriand committed:
Issue 534 (#554)

* Travis test

* Change CXX11 ABI flag

* Try to build dependencies on Travis

* Changed clone url

* Changed cmake source dir

* Try llvm 6.0.1

* Revert "Try llvm 6.0.1"

This reverts commit 40b87d3ed1d16573817e48c86e5a128ba2d4a24d.

* remove extraneous mkdir

* Set CXX ABI to 1

* Try only change ABI flag

* Add travis tests for the branch

* Update .travis.yml

Disable CXX11 ABI

* Change static linker flags

* Revert "Change static linker flags"

This reverts commit 6442d68252d4334e2d926c9c0eab31a741f08b00.

* recompile everything

* Recompile everything using sudo

* Clean objective files

* Revert cleaning object files

* Tried recompiling the debs

* Fixed deb file name

* Try disabling CXX14

* go back to original debs

* Try compiling debs with C++11 ABI disabled

* Recompiled debs with C++11 ABI enabled

faye-u committed:
State saving (#556)

* Started working on state saving

* continued state saving

* trying to pause

* Can pause sim

* Serialization functions for LLVMModelData and rrSparse

* State saving for LLVMModelDataSymbols

* Temporary testing code

* Save LLVM module

* Started cleaning up state saving interface

* Add save and load states in Roadrunner

* Continued work on state saving

* Fix merge conflict

* Transfer module pointer from generation to model

* Switched to IR because of corrupt bitcode error

* cleaning

* readded comment accidentally removed

* Fixed LLVM related crash

* Try to fix tests 28, 173, 269, 1120

* Add modelGeneratorOpt parameter to state loading functions

* Save LibStructural by sbml for now, later save by matrix

* Fix bug where model could only be saved after loading from SBML

* Removed reset in load method

* Fix tests 28 173 269

* Fix test 1120

* cleanup

* Save integrator parameters

* Remove createGlobalMappingFunction

* Fix column number in simulation result

* Save column names

* Added some comments

* Changed to LLVM Bitcode for saving

* Fix includes

* Add file IO error handling

* Try disabling USE_TR1_CXX_NS for travis

* Add double vector to rr::Variant from absolute-tolerances

* Add double vector case to loadBinary and saveBinary

* Remove testing code from rrc_api.cpp

* Try using rr::Variant::convert for double vector

* Add magic and version number

* Hide LLVMExecutableModel from RoadRunner

* Switch to loadBinary and saveBinary in LLVMModelData

* Clear map in loadBinary

* Remove extraneous includes

* Start to work

* Enable to reload in the middle of simulation

* Document saveBinary and loadBinary

* Change saveBinary documentation

* Save steady state solver params

* Prototype

* Merge issues

* Delete duplications

* Documentation for saveState and loadState

* Invalid option handling

* Store object file instead of LLVM IR

* Fix exceptions

* Use PIC relocation

* Get TargetMachine from ExecutableModel

* Global mappings

* Don't attempt to set maximum_adams_order

* Put back simulate function

* Remove debugging code

* Added C api functions for save and load state

* Fix resaving bug

* Prevent saving an empty RoadRunner instance

* Comment new save state

* Use check_model

* Added save and load state to C API header

* Generate object file instead of module to optimize state saving

* Start state saving test suite

* Continue state saving test suite

* restart integrator in loadState

* Continue state saving test suite

* Save pending events

* Continue test suite

* Better naming scheme for state saving tests

* Remove extraneous parameter from RunStateSavingTest

* Rename state saving test folder to match suite name

* Move state saving tests to cxx tests

* Test breaking travis build with cxx tests

* Unbreak travis build

* Get sbml_test_suite.cpp from model-editing

* State saving C API Tests

* Update documentation

* Hide rr::ExecutableModel::saveState from python interface

andrewbriand committed:
Absolute tolerances (#555)

Add the ability to set absolute tolerances for individual species

* Modified rr::Variant to allow it to store std::vector<double>

* Add vector absolute tolerance in CVOD integrator

* Added to rr::Variant documentation

* CVODEIntegrator

* Set absolute tolerance based on concentration seperately

* Fix vector script out of range

* Modify loading setting files

* Fix illeagal input when the tolerance is equal to 0

* Get the compartment volume for each species

* Convert python double list to variant

* Add documentation for CVODEIntegrator

* Add documentation for concentration tolerance

* Add documentation for concentration tolerance

* Add documentation for concentration tolerance

* Add API to set tolerance for individual species

Need to update roadrunner.py as well

* Convert double vector variant to list in python

* Modify API to set concentration tolerance

* Add error handling for incorrect vector size

* Error handling for concentration tolerance

* Error handling when set individual tolerance

* Enable conversion from int or float input to double

* Error handling for wrong type

* Add getter for concentration tolerance

* Documentation for new APIs

* Documentation for new APIs

* Move new functions from Solver to Integrator

* Documentation for new APIs

* Parse double vector differnetly

* Fix not freed memory block

* Try to fix errors on Travis

* Try to fix errors on Travis

* Still try to fix errors....

* Keep fixing errors

* Fix errors on Travis

* Ignore files in .vs/

* Fix incorrect size for tolerance vector when setting scalar concentration tolerance

* Modify API to set concentration for individual variable

* Delete allocated array

* Fix incorrect size for tolerance vector when setting vector concentration tolerances

* Update python doc for vector absolute tolerances

* Modify getConcentrationTolerance based on the correct vector layout

* Update API in Integrator

* C API to set double array

* Fix incorrect return type

* C API to set concentration tolerance

* C API to set individual tolerance

* Modify C API to return double array and length in arguments

* Add tests for integrator

* Add python testing methods for new API

* Fix minor bug in tester

* Remove unnecessary test

* Fix incorrect format

* Add testID parameter

* Fix test1

* Fix typo in test functions

* Remove .vs folder

faye-u committed:
Model editing (#558)

Add an API for editing models on the fly

* Copy old values by modifying data block directly

* More test cases

* more test cases

* Replace the confusing documentation in ModelData

* Revert to previous method to copy old values

* Copy old values for rate rule

* Only one assignment per variable per event

* Put loop back

* Test removeEvent and removeEventAssignments

* Add delay 2

* Remove duplicated columns for species with rate rules from the deault selection list

* Check and remove all parameter without any assignments

* More tests

* Fix overwriting old values of rateRule

* Check model is not NULL before deleting

* Try to fix ADD_PRIORITY_2

* Copy initial values and currt values seperately

* Fix test PAUSE_1

* Fix test PAUSE_3

* Word on ADD_PRIORITY_2, uncomment

* Fix nullptr when removing parameters recursively

* Fix test ADD_PRIORITY_2

* Add C API for addTrigger

* Remove extraneous output

* Add python doc for remove/add parameter

* Add python doc for add/remove compartment

* Readd species test

* Set required attributes to default values

* Python doc for add/remove rules

* Check for null kinetic law

* Test recursive parameter removal

* Add python doc for editing events

* Fix typos

* Remove param recurse test 2

* Readd check for null kinetic law

* Remove event assignment test 2

* Clean up testing code

* Move tests to C++ folder, rename tests to match code

* Finish renaming C++ tests

* Uncomment SBML tests in c++ tests

* Try to run cxx tests on travis

* Change travis cxx test arguments

* Fix source dir

* Remove cxx tests from c test file

* Model Editing C API tests

* Start to work on Python test suite

* Fix non-declared function

* Call editing function after initialization

* Making a model from scratch tests

* Remove test reset

* Add test method for addReaction

* More from scratch tests

* AddReaction Python test 1

* Fix AddReaction1

* Remove tests from SBML-l2v4 that don't have an l2v4 version

* Add more test methods

* Set stiff to false by default in cxx tests

* AddSpecies2

* Add evnet assignment

* Add methods to test remove functions

* Fix unknown tests

* Fix failed concentration tests

* RemoveSpecies test

* RemoveCompartment test

* RemoveParameter test

* RemoveRules test

* Fix cycle in assignment rules

* Fix RemoveParameter test

* Fix incorrect test file format

* Fix wrong tag

* SetKineticLaw test

* Make resources modelData and symbols private

* Update comments

* Fix potential nullptr when reaction doew not has a kinetic law

* Make ModelGeneratorContext::module private

* Remove unused constructor

* Not check global parameter when removing events

* Remove setKineticLaw test

* Remove unused local

* COMMENT THE CORRECT LINE THIS TIME, HOPEFULLY

* Remove debug call

* Remove extraneous methods

* Uncomment ModelGeneratorContext dtor

* Remove unused constructor

* Remove string vector from Variant

* Test if the parameters could be removed recursively

* Test remove recursively

* Use current SBML when setting conserved moiety analysis

* Enable validation in load options

* Fix building issues

* Fix illegal model

* Set constant and boundary condition for species

* Try to fix test errors in removeRules1

* Set reversible for reaction

* Set constant attribute for all three variables

* Set persistent and initValue for trigger

* Fix wrong target to remove rules

* Still try to fix removeRules test failure

* Validate SBML model after model editing tests

* Use SBMLDocument default constructor

* Remove the failed test

* Don't check modeling practices in model-editing tests, fix PAUSE_6

* Infer modifiers from kinetic law

* Fix typo in getSpeciesIdsFromAST declaration

* Return number of failed tests for cxx_api_tests

* Set sbml version appropriately is from scratch tests

* Not overwrite the initial value, HOPEFULLY

* Write the old values directly to data block in case of READ_ONLY mode

* Fixing moiety model generation

* Regenerate for conserved moieties

* fix typo

* readd document based constructor

* Fix document based ModelGeneratorContext constructor

* Revert the loading line to see if tests passed

* Check if index is out of range when copying the old initial values

* Regenerate properly in READD_SPECIES

* Fix some readd species tests

* Don't run test 21 in readd_species, regenerate in readd_compartments

* Setter for hasOnlySubstanceUnits

* Setters for initial amount and concentration

* Fix PAUSE_8

* set has only substance units in READD_SPECIES

* Call regenerate for conserved moiety analysis

* Add testing methods for new API

* Fix wrong reaction order

* Fix wrong type in setting init amount/concentration

* C API setters for species

* C API new setters for reaction and trigger

* Always check if index is in the range when copying old values

* Update regenerate

* Set initAmount by default when adding species

* setPersistent tests

* Check if sid is a valid variable when adding rules

* Check for duplicated rules

* setConstant tests

* No multiple assignments for the same variable in one event

* Fix typo

* Check valid variable in addEventAssignment

* No multiple assignment rules for the same varaible in one event

* setHasOnlySubstanceUnits tests

* Remove unnecessary rules

* Set initial concentration test

* Add sbml tag

* Fix setters for initial values of species

* Fix typo

* Remove incorrect reset call

* Set initial concentrations and amounts

* Recover the initial initial value after the assignment rule is removed

* Change the order to recover the initial value

* Fix missing initial value

* More tests

* Uncomment tests

* Recover the updated initial values after regen

* Fix function not in the scope

* Recover initial concentration

* C API Tests for new setter methods

* Don't do readd species on models with events and rules

* Remove debug code

* Update initial concentration in removeRules

* Reset after setting trigger initial value

* Reset after setTriggerInitialValue in C Tests

* Copy the old time to new model

* Remove rules python test

* Fix remove rules test

* Remove multiple rules

* Test for assignment rules

* Add assignment rule for global parameter

* Evaluate trigger initial value

* Remove reset int setTriggerInitialValue tests

* Restart with current time

* Don't apply events at T0 whose initialValue has been set to true

* Don't trigger T0 events 3 times

* Only reset once under default options

* Not test for assignment rules

* Fix cxx sbml tests

* Uncomment tests

* Reset in setTriggerInitialValue if simulate hasn't been called

* Combine reset options

* Set validation flag

* Turn on validation during regeneration

* Add new key in Config

* Remove unnecessary function

* Add a flag to recover initial values after removing assignment rules

* Add a new function exposed for validation

* Remove debug statements

* Update python test for removeRules

* doc -> _doc to avoid confusion

* C API for validation function

* Turn off unit validation for now

* Maintain the species order

* Not test for assignment rule

* Readd events

* remove rules tests

* Add and remove initial assignment

* Check for existence of assignment rule and other initial assignment

* Recover initial values after removing rate rules

* Reset after editing events

* Move else branch to the correct potision

* REMOVE_RATE_RULE_4

* Integrate reset in to regenerate

* Update header file

* Recover initial values after removing initial assignment

* Only reset variables from initalAssignments at t=0

* REMOVE_INITIAL_ASSIGNMENT_1

* remove extra method calls in constructor

* Initial assignment takes effect

* Fix compiler errors

* Fix .travis.yml

andrewbriand committed:
stream read and write to loadBinary and saveBinary in rrSparse

andrewbriand committed:
Transfer absolute tolerance vector from old to new model upon regeneration

andrewbriand committed:
Fixed bug in RoadRunner::regenerate and added test

andrewbriand committed:
Add RETAIN_ABSOLUTE_TOLERANCES_2 test

andrewbriand committed:
Test for retained absolute tolerances after loading state

andrewbriand committed:
fix bug in interaction between model editing and state saving, add tests

andrewbriand committed:
More tests

andrewbriand committed:
Fix illegal write in regenerate

andrewbriand committed:
Fix some memory leaks

andrewbriand committed:
Fix more mem leaks

andrewbriand committed:
properly initialize model and document

andrewbriand committed:
Fix double deletion

andrewbriand committed:
Fix memory leak

andrewbriand committed:
Use unique_ptr to executable model

andrewbriand committed:
merge

andrewbriand committed:
Fix compiler error

andrewbriand committed:
Fix error in regenerate

andrewbriand committed:
Use unique_ptr for doc

andrewbriand committed:
Refactor ModelGeneratorContext::stealThePeach

andrewbriand committed:
Fix memory leaks

andrewbriand committed:
Merge branch 'llvm-6' of https://github.com/sys-bio/roadrunner into llvm-6

andrewbriand committed:
Fix memory leak in test suite

andrewbriand committed:
Refactor CVODEIntegrator::variableStepPostEventState

andrewbriand committed:
Test for issue 562

andrewbriand committed:
Update properties on relevant model editing actions

andrewbriand commented on source/llvm/LLVMModelGenerator.h:
```diff
@@ -31,11 +34,16 @@ class RR_DECLSPEC LLVMModelGenerator
     static rr::ExecutableModel *createModel(const std::string& sbml, uint options);
 
+	/**
+	 * Regenerate an executable model from an existing sbml document
```
This documentation should be more detailed, what exactly is regeneration in this context?

andrewbriand commented on source/llvm/ModelGeneratorContext.cpp:
```diff
@@ -152,24 +159,32 @@ ModelGeneratorContext::ModelGeneratorContext(std::string const &sbml,
         // initialize LLVM
         // TODO check result
         InitializeNativeTarget();
+		InitializeNativeTargetAsmPrinter();
```
Do these need to be here?

andrewbriand commented on source/llvm/ModelGeneratorContext.cpp:
```diff
@@ -253,43 +264,48 @@ ModelGeneratorContext::ModelGeneratorContext(libsbml::SBMLDocument const *doc,
         // TODO check result
         InitializeNativeTarget();
+		InitializeNativeTargetAsmPrinter();
+		InitializeNativeTargetAsmParser();
```
Do these need to be here?

andrewbriand commented on source/llvm/ModelGeneratorContext.h:
```diff
@@ -187,11 +190,17 @@ class ModelGeneratorContext
 
+public:
+
+    llvm::Module* module;
```
Does this need to be public?

andrewbriand commented on wrappers/Python/roadrunner/roadrunner.i:
```diff
@@ -692,6 +693,13 @@ PyObject *Integrator_NewPythonObj(rr::Integrator* i) {
 %rename (_getConservedMoietyAnalysis) rr::RoadRunner::getConservedMoietyAnalysis();
 %rename (_setConservedMoietyAnalysis) rr::RoadRunner::setConservedMoietyAnalysis(bool);
 
+//Model editing proxies so that we can regenerate properties each time we regenerate the model
```
Should we regenerate properties when removing variables also?

andrewbriand committed:
Don't need to initialize asm parser and printer

andrewbriand committed:
Revert

andrewbriand committed:
Improve documentation, make ModelGeneratorContext::module private

andrewbriand committed:
Merge branch 'develop' into llvm-6

andrewbriand committed:
Bump version to 1.6.0

andrewbriand committed:
Fix issue with loading

andrewbriand committed:
Improve docs

andrewbriand committed:
Initialize LLVM target in RoadRunner ctor, add test

andrewbriand committed:
Issue 566

andrewbriand committed:
Add missing test model

andrewbriand committed:
Add tests for issue 566

andrewbriand committed:
Fix bug, add tests

andrewbriand committed:
Add fix to work with libSBML-5.18.1

andrewbriand committed:
Ast distrib (#569)

* Add support for AST_DISTRIB

* Refactored the code for ASTNodeCodeGen to take ModelGenContext and ModelData

* Support truncation for AST_DISTRIB

* Make compatible with llvm-6

andrewbriand committed:
Add option to remove unused parameters
